### ISSUE TITLE
Extract Class

### ISSUE THREAD

**Issue by retailcoder** [rubberduck-vba/Rubberduck#892]
We need a refactoring that helps VBA programmers to refactor procedural code into object-oriented code. Extracting methods is step one - extracting closely related methods into a class is step two.

#178 / Extract Interface would be step 3.

This refactoring works off the active code pane - it doesn't matter that we're in a standard, class, or form module. It displays a dialog that lists all members; user selects which members to move into the new class, enters a name for that new class, and Rubberduck...

 - Adds a new class module, names it as specified by the user
 - Adds a new private, self-initialized (`As New Foobar`) field in the original module, of the type of the extracted class.
 - Adds the selected members as public members of the extracted class
 - Replaces method calls with member calls (on the new field instance) in the original module

Hmm, perhaps the dialog also needs a place to specify a name for the private field.

**Comment by Hosch250** [comment]
How would the calls change?  Would they all just change from `Foo(param1, param2)` to `newField.Foo(param1, param2)`?

**Comment by Hosch250** [comment]
Also, what members are applicable?  Fields, properties, methods?  What about events?  What about interface implementation members?

**Comment by retailcoder** [comment]
Anything in the module body: functions, procedures and properties.

**Comment by retailcoder** [comment]
Exclude interface members and event handlers... keep it simple :-)

**Comment by retailcoder** [comment]
Call sites will need a new field if there's more than 1 call, or a new local if there's only 1 (make the object as tightly-scoped as possible) - you'll need to prompt for a name and replace calls with member calls on the new object.

**Comment by Hosch250** [comment]
Could you give me a few compilable examples of the different syntaxes I need to use?  Also, I am supposed to delete the existing items, correct?  (If so, the rewriter will be perfect - remove them from the entire module, then rewrite the entire module in one blow.)

**Comment by retailcoder** [comment]
Yes, the extracted members should be removed from their original location, and call sites rewritten so that the code still compiles and runs just like before the refactoring.

If an extracted method has a dependency on a member that isn't selected, that member should be automatically selected - you don't want to introduce circular dependencies and tangle the code into a mess.... the goal of refactoring is to clean up code!

**Comment by Hosch250** [comment]
This could quickly become a mess.  Selecting enough members will move the entire class over.  One solution could be to copy unselected dependencies and move selected dependencies.

**Comment by retailcoder** [comment]
I'd like the thing to work just like the way ReSharper has it.. ideally :--)

**Comment by Hosch250** [comment]
I looked at R#'s version, and it sounds like it is essentially my suggestion, with a few other refinements.  However, my classes are beginning to go up, so I am not sure if I will be able to complete this in time.  Feel free to take over where I have left off if I do not complete it.

**Comment by retailcoder** [comment]
I'm starting to think this one might be a little too complex and time-consuming to release with 2.0.. how about we keep it for a later release, when the dialogs are in XAML? The UI will be easier to get right, and we'll have time to tweak the parser API to intake a modified token stream.

**Comment by Hosch250** [comment]
Fine by me.  I'm thinking I want to take this week pretty much off before school gets in full swing again anyway.  I won't delete my branch, though.  I've done some good work here.
